modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/IntrospectionUtils.java

```diff
@@ -536,24 +541,87 @@ public static String getSourceName(Class<? extends Source> sourceType) {
    * @param extensionModel a {@link ExtensionModel}
    * @return a non {@code null} {@link Set}
    */
-  public static Set<Class<?>> getParameterClasses(ExtensionModel extensionModel) {
-    ImmutableSet.Builder<Class<?>> parameterClasses = ImmutableSet.builder();
+  public static Set<Class<?>> getParameterClasses(ExtensionModel extensionModel, ClassLoader extensionClassLoader) {
+    Set<Class<?>> parameterClasses = new HashSet<>();
     new ExtensionWalker() {
 
       @Override
       public void onParameter(ParameterizedModel owner, ParameterModel model) {
-        parameterClasses.add(getType(model.getType()));
+        collectRelativeClasses(model.getType(), parameterClasses, extensionClassLoader);
       }
     }.walk(extensionModel);
 
-    return parameterClasses.build();
+    return parameterClasses;
   }
 
+
+  /**
+   * Given a {@link MetadataType} it adds all the {@link Class} that are related from that type. This includes generics of an
+   * {@link ArrayType}, key and value of an {@link DictionaryType} and classes from the fields of {@link ObjectType}.
+   *
+   * @param type {@link MetadataType} to inspect
+   * @param relativeClasses {@link Set} were the classes should be added
+   * @param extensionClassLoader extension class loader
+   */
+  public static void collectRelativeClasses(MetadataType type, Set<Class<?>> relativeClasses, ClassLoader extensionClassLoader) {
```

Comment:
should not receive the set... should return a new set instead. If you move this method into an utils, then it should have utils semantics, not speculate with how it should be used